### ISSUE TITLE
Decompiler executability support for aarch64/arm64 (linux, and possibly macOS)

### ISSUE THREAD

**Issue by MrKOSMOS** [NationalSecurityAgency/ghidra#3622]
**The first title (sorry it's my first feature request)**

Arm-powered linux devices are fairly ubiquitous now, from SBCs like the Raspberry Pi to almost every smartphone/tablet, honorable mentions being the new Macbooks. Even though they are mostly unfit to be daily-drivers, many people own at least one of these, and given the chance would carry a much lighter/cheaper device to work/to a café if able to run their preferred software. 
Even Android devices are now capable of running a windows manager and a suitable desptop environment, through apps such as Termux and UserLAnd through VNC.

**Running Ghidra on these devices**
Now, there is a current JDK release for aarch64. Heck, in fact, Ghidra even launches with no errors, and disassembles binaries with no problem. Errors only arise when attempting to **decompile** the binaries, as shown in the screenshots below. And at least to me, the decompilation results are key to understanding a binary. Also, given Ghidra's amazing flexibility and power, it is my go-to disassembler and decompiler for retro games and modern discontinued software alike.
I would honestly try to correct this myself and rebuild the decompiler executable for arm64, but the number of source code changes needed in order to include arm64/aarch64 quickly turned out to be too great to find and make on my own. 
However, as far as I see, there shouldn't be any massive hurdles to adding support, little more than including arm64/aarch64 within the Linux options on every decompiler source file, with macOS soon to follow.
I hope you see my point, or at least would like to run Ghidra on your smartphones as a simple gag.

**Screenshots**
And finally, the screenshots. The fact that I'm running a VNC window means it's impossible to copy text directly, but I may be able to copy everything into a text file and add it here later if necessary, and hopefully export it to main storage even though rootless Android 11 heavily limits storage access in applications
![Screenshot_20211109-213247_VNC Viewer](https://user-images.githubusercontent.com/24480840/141115941-86889f1d-abda-4676-9b44-fa068525b6a6.jpg)
![Screenshot_20211109-213315_VNC Viewer](https://user-images.githubusercontent.com/24480840/141115948-74e2a35a-0f80-42d5-9127-c20439bec13a.jpg)
![Screenshot_20211109-213504_VNC Viewer](https://user-images.githubusercontent.com/24480840/141115956-2dcd5dad-2506-4768-8d3f-dfdcf9f8b886.jpg)
.

**Comment by ryanmkurtz** [comment]
The master branch supports building aarch64 natives on both Linux and macOS. The next version of Ghidra we will release will also support building/rebuilding the natives contained in the release for these platforms too.

**Comment by MrKOSMOS** [comment]
Oh okay thank you. I guess it's... Solved?

**Comment by astrelsky** [comment]
> Oh okay thank you. I guess it's... Solved?

We'll find out one way or another :rofl:. The only thing worse then an edge-case failure is when your test cases were, in fact, the edge cases.

**Comment by ryanmkurtz** [comment]
> Oh okay thank you. I guess it's... Solved?

Yes, solved by #3197 and #3387

**Comment by MrKOSMOS** [comment]
> > Oh okay thank you. I guess it's... Solved?
> 
> We'll find out one way or another 🤣. The only thing worse then an edge-case failure is when your test cases were, in fact, the edge cases.

I'll make sure to test it once it comes out then!

**Comment by ryanmkurtz** [comment]
Sounds good. You can also get it up and running right now if you clone the repo and build the master branch on your aarch64 platform. The instructions are in the README.md file.

**Comment by MrKOSMOS** [comment]
Alright I might try that.

**Comment by MrKOSMOS** [comment]
Building on Termux gives an error. I let it run twice and logged the results (attached file).
Fortunately, it seems to be a Java compiler error. Sadly, it's somewhere else I need to file a report...
So I am forced to build using UserLAnd, which sadly seems to only use one thread (?) making the whole process veeeeeery slow (10% after 30 minutes)... But it works.
It must be an issue with the JDK 17, since the only available JDK package in UserLAnd is default-jdk which installs JDK 11. Termux sadly doesn't support installing external packages, and the only available package is said JDK 17.
Other than that, it just throws a few "deprecated feature soon to be unsupported in Gradle 8.0", so I might check for an issue report and create one of my own if there is none.
[errorlog.txt](https://github.com/NationalSecurityAgency/ghidra/files/7530193/errorlog.txt)

**Comment by ryanmkurtz** [comment]
Ghidra not compiling with JDK 17 is a known issue. I've reported the problem to Oracle and I expect it to be fixed in their upcoming 17.0.2 release. The version of Java that Ghidra officially supports is JDK 11.

**Comment by MrKOSMOS** [comment]
Okay, compiled fine on UserLAnd and seems to work great, no warnings, errors or crashes while decompiling random PD Game Gear games. And the decompiler WORKS!
Also gradle works amazingly, never have I ever seen a compiler return to where it was after an error (missing dependencies, my bad, the UserLAnd distro seems to be pretty barebones) or a system crash (also my bad, Android kills idle apps in background).
Or maybe I'm just too used to Make. Ah, the memories of leaving RetroPie updating for a whole day... 
Anyway, hopefully most compilation warnings will be solved by Gradle 8.0.
Thanks y'all.

**Comment by astrelsky** [comment]
@MrKOSMOS may you clarify what you got it running on? Just curious.

**Comment by MrKOSMOS** [comment]
@astrelsky I'm not completely sure myself, but I'll give you the rundown.
So I have an octa-core ARM Android 11 tablet, and I pretty much installed what seems to be a "compatibility layer" app called UserLAnd, which I think uses the native Linux capabilities of Android OS and complements it with its own standard Linux libraries in order to run standard (aarch64) applications.
It has various "distro" and "application" options, including out-of-the-box a few distros, desktop environments and GNU applications which download on-demand, of which I chose the "lxde Desktop". Hopefully it was the most minimal option. In fact, I had to manually install dialog, apt-utils, gcc and g++ in order to compile Ghidra, since it didn't include them. The installable Gradle version is 4.4.1, which is very old, but at least I could install the latest one from the Gradle website. Only installed those packages to prevent my free storage from instantly dwindling into the void...
Oh also, it uses VNC (there's an android app) for display, since I really just want to run Ghidra.
Some of the other options actually use/support ssh, including the distros, making it similar to Termux. I didn't use Termux though, even though it's easier to get files to and from the app sandbox storage, because:
-only includes JDK 17, which bugs out building Ghidra, and doesn't allow installing .rpm or .deb packages;
-the VNC host in Termux requires command line input to start,  while in UserLAnd it only needs the push of a button (yes I'm lazy)
And that's all.
